src/test/jtx/subcases.h

```diff
@@ -0,0 +1,131 @@
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_JTX_SUBCASES_H_INCLUDED
+#define RIPPLE_TEST_JTX_SUBCASES_H_INCLUDED
+
+#include <xrpl/beast/unit_test/suite.h>
+
+#include <cstdint>
+#include <functional>
+
+namespace subcases {
+
+constexpr std::size_t MAXIMUM_SUBCASE_DEPTH = 10;
+
+/**
+ * This short library implements a pattern found in doctest and Catch:
+ *
+ * TEST_CASE(testName) {
+ *     // setup
+ *     SUBCASE("one") {
+ *         // actions and assertions
+ *     }
+ *     SUBCASE("two") {
+ *         // actions and assertions
+ *     }
+ *     SUBCASE("three") {
+ *         // actions and assertions
+ *     }
+ *     // assertions before teardown
+ * }
+ *
+ * EXECUTE(testName);
+ *
+ * In short:
+ *
+ * - Top-level test cases are declared with `TEST_CASE(name)`.
+ *   The name must be a legal identifier.
+ *   It will become the name of a function.
+ * - Subcases are declared with `SUBCASE("description")`.
+ *   Descriptions do not need to be unique.
+ * - Test cases are executed with `EXECUTE(name)`,
+ *   where `name` is the one that was passed to `TEST_CASE`.
+ *   When executing a test case, it will loop,
+ *   executing exactly one leaf subcase in each pass,
+ *   until all subcases have executed.
+ *   The top-level test case is considered a subcase too.
+ *
+ * This lets test authors easily share common setup among multiple subcases.
+ * Subcases can be nested up to `MAXIMUM_SUBCASE_DEPTH`.
+ */
+
+struct Context
+{
+    beast::unit_test::suite& suite;
+    // The number of subcases to skip at each level to reach the next subcase.
+    std::uint8_t skip[MAXIMUM_SUBCASE_DEPTH] = {0};
+    // The subcase names at each level.
+    char const* names[MAXIMUM_SUBCASE_DEPTH] = {""};
+    // The current level.
+    std::uint8_t level = 0;
+    // The maximum depth at which we entered a subcase.
+    std::uint8_t entered = 0;
+    // The number of subcases we skipped on this or deeper levels
+    // since entering a subcase.
+    std::uint8_t skipped = 0;
+
+    std::string
+    name() const
+    {
+        std::string n;
+        for (auto i = 0; i <= level; ++i)
+        {
+            if (i != 0)
+            {
+                n += " > ";
+            }
+            n += names[i];
+        }
+        return n;
+    }
+
+    void
+    lap()
+    {
+        level = 0;
+        entered = 0;
+        skipped = 0;
+    }
+};
+
+struct Subcase
+{
+    Context& _;
```

Comment:
please do not use `_` as a variable or data member name; C++ is about to get special meaning for it https://wg21.link/P2169

Reply:
... already in C++ draft https://eel.is/c++draft/basic.scope.scope#5